### PR TITLE
Cleanup federation phase 2

### PR DESCRIPTION
**Pull request by VicDeo**
## Description
Follow up on https://github.com/owncloud/core/pull/32086
- Separates input params validation and handling external shares
- Increases code coverage
- Decreases C&P at some degree

## Related Issue
https://github.com/owncloud/enterprise/issues/2564

## Motivation and Context
Extract business logic from controller, make it reusable.

## How Has This Been Tested?
Testing in progress... basic tests are:
#### Case 1 :heavy_check_mark: 
- [owner] create a federated share with share permision
- [recipient] accept it

expectations:
[x] owner see that file is shared
[x] recipient see a file and can reshare 

- [owner] change share permissions

expectations: 
[x] recipient can NOT reshare any more

- [recipient] unshare it
expectations: 
[x] owner see that file is NOT shared

#### Case 2 :heavy_check_mark: 
- [owner] create a federated share with sharing perms
- [recipient] accept it
- [recipient] reshare it with another user

expectations: 
[x] another user see a reshare

- [owner] revokes a share

expectations: 
[x] recipient can NOT see the file any more
[x] another user can NOT see a reshare

#### Case 3 :heavy_check_mark: 
- [owner] create a federated share
- [recipient] decline it

expectations: 
[x] owner see that file is NOT shared

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Technical debt
- [x] Tests

## Checklist:
- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

### CODE REVIEW

File: apps/federatedfilesharing/lib/Controller/RequestHandlerController.php [hunk 1]
```diff
@@ -113,112 +102,79 @@ public function __construct($appName,
...
-			$shareWith
-		);
-
 		try {
-			$externalManager->addShare($remote, $token, '', $name, $owner, false, $shareWith, $remoteId);
-			$shareId = \OC::$server->getDatabaseConnection()->lastInsertId('*PREFIX*share_external');
-			if ($ownerFederatedId === null) {
-				$ownerFederatedId = $owner . '@' . $this->cleanupRemote($remote);
+			$this->assertIncomingSharingEnabled();
+			$remote = $this->request->getParam('remote', null);
```
Comment (PVince81) [inline apps/federatedfilesharing/lib/Controller/RequestHandlerController.php]: shouldn't we specify those as method params instead?

File: apps/federatedfilesharing/lib/Controller/RequestHandlerController.php [hunk 2]
```diff
@@ -235,44 +191,50 @@ public function createShare() {
...
-			) {
+			if ($this->addressHandler->compareAddresses($user, $remote, $owner, $currentServer)) {
 				return new Result(null, Http::STATUS_FORBIDDEN);
 			}
 
-			// check if re-sharing is allowed
-			if (!$share->getPermissions() | ~Constants::PERMISSION_SHARE) {
+			$reSharingAllowed = $share->getPermissions() & Constants::PERMISSION_SHARE;
+			if (!$reSharingAllowed) {
 				return new Result(null, Http::STATUS_BAD_REQUEST);
```
Comment (PVince81) [inline apps/federatedfilesharing/lib/Controller/RequestHandlerController.php]: should we add a message here for clarity ?

File: apps/federatedfilesharing/lib/FedShareManager.php [hunk 3]
```diff
@@ -22,54 +22,185 @@
...
+	 * @param string $name
+	 * @param int $ownerFederatedId
+	 * @param int $sharedByFederatedId
+	 * @param string $sharedBy
+	 * @param string $token
+	 *
+	 * @return void
+	 */
+	public function createShare($shareWith,
+								$remote,
```
Comment (PVince81) [inline apps/federatedfilesharing/lib/FedShareManager.php]: I guess there's no way currently to pack all this into a nice object ? array doesn't sound that attractive and IShare is likely not suitable.

Just thinking out loud... let's leave this for now

File: apps/federatedfilesharing/lib/FedShareManager.php [hunk 4]
```diff
@@ -119,6 +252,81 @@ public function declineShare(IShare $share) {
...
 
+	/**
+	 * Unshare an item
+	 *
+	 * @param int $id
+	 * @param string $token
+	 *
+	 * @return void
+	 */
+	public function unshare($id, $token) {
```
Comment (PVince81) [inline apps/federatedfilesharing/lib/FedShareManager.php]: `reShare` vs `unshare`, why not `unShare` then ? I guess you just copied the inconsistent names from the original handler :-D

File: apps/federatedfilesharing/lib/FederatedShareProvider.php [hunk 5]
```diff
@@ -956,4 +959,70 @@ public function updateForRecipient(IShare $share, $recipient) {
...
+	 * @param $token
+	 * @param $name
+	 * @param $owner
+	 * @param $shareWith
+	 * @param $remoteId
+	 *
+	 * @return int
+	 */
+	public function addShare($remote, $token, $name, $owner, $shareWith, $remoteId) {
+		\OC_Util::setupFS($shareWith);
```
Comment (PVince81) [inline apps/federatedfilesharing/lib/FederatedShareProvider.php]: assuming $shareWith was already validated before as being an existing user ?

I checked setupFS and it seems it wouldn't throw an exception.

If this is from the old code as is, feel free to leave it for now

**Review by PVince81 — APPROVED** [review]
:+1: looks good to me

see comments for optional changes

File: apps/federatedfilesharing/lib/FederatedShareProvider.php (hunk 5 above)
Comment (VicDeo) [reply apps/federatedfilesharing/lib/FederatedShareProvider.php]: yes, in controller  https://github.com/owncloud/core/pull/32113/files#diff-47cd1c326a2fae00c6f0e1e088c994bbR141

File: apps/federatedfilesharing/lib/Controller/RequestHandlerController.php (hunk 1 above)
Comment (VicDeo) [reply apps/federatedfilesharing/lib/Controller/RequestHandlerController.php]: probably yes, but I did not test it this way

File: apps/federatedfilesharing/lib/Controller/RequestHandlerController.php (hunk 2 above)
Comment (VicDeo) [reply apps/federatedfilesharing/lib/Controller/RequestHandlerController.php]: Validate all error statuses and messages against the spec is another topic that I don't want to start in this PR 
It has enough modifications to be barely readable :)

File: apps/federatedfilesharing/lib/FedShareManager.php (hunk 3 above)
Comment (VicDeo) [reply apps/federatedfilesharing/lib/FedShareManager.php]: yes, it can be the IShare  object with some additional params

**Comment by VicDeo** [comment]
@PVince81 let's merge it for now, I'll put a checkbox into the task to address the comments in the next PR.

**Comment by VicDeo** [comment]
Stable10: https://github.com/owncloud/core/pull/33027

**Pull request merged**